src/iocore/cache/RamCacheContainer.cc

```diff
@@ -0,0 +1,322 @@
+  // returns 1 on found/stored, 0 on not found/stored, if provided auxkey1 and auxkey2 must match
+  int     get(CryptoHash *key, Ptr<IOBufferData> *ret_data, uint64_t auxkey = 0) override;
+  int     put(CryptoHash *key, IOBufferData *data, uint32_t len, bool copy = false, uint64_t auxkey = 0) override;
+  int     fixup(const CryptoHash *key, uint64_t old_auxkey, uint64_t new_auxkey) override;
+  int64_t size() const override;
+  void    init(int64_t max_bytes_, StripeSM *stripe_) override;
+  RamCacheContainer();
+  ~RamCacheContainer() override;
+};
+
+extern int cache_config_ram_cache_algorithm;
+extern int cache_config_ram_cache_numa_duplicate;
+
+RamCacheContainer::RamCacheContainer() : caches(numa_max_node() + 1, nullptr) {}
+
+RamCacheContainer::~RamCacheContainer()
+{
+  for (auto c : caches)
+    delete c;
+}
+
+RamCache *
+new_RamCacheFromConfig()
+{
+  switch (cache_config_ram_cache_algorithm) {
+  default:
+  case RAM_CACHE_ALGORITHM_CLFUS:
+    return new_RamCacheCLFUS();
+  case RAM_CACHE_ALGORITHM_LRU:
+    return new_RamCacheLRU();
+  }
+}
+
+RamCache *
+new_RamCacheContainer()
+{
+  return new RamCacheContainer;
+}
+
+void
+RamCacheContainer::init_one_cache()
+{
+  unsigned int my_node = 0;
+  getcpu(nullptr, &my_node);
+  if (caches[my_node]) {
+    ink_error("Attempt to double-init duplicated cache!");
+  } else {
+    RamCache *cache = new_RamCacheFromConfig();
+    cache->init(max_bytes, stripe);
+    caches[my_node] = cache;
+  }
+}
+
+RamCache *
+RamCacheContainer::get_cache(unsigned int my_node, unsigned int node)
+{
+  return caches[node];
+}
+
+static void *
+ram_cache_container_thread_init_func(void *arg)
+{
+  ((RamCacheContainer *)arg)->init_one_cache();
+  return nullptr;
+}
+
+void
+RamCacheContainer::init(int64_t max_bytes_, StripeSM *stripe_)
+{
+  max_bytes   = max_bytes_;
+  stripe      = stripe_;
+  init_called = true;
+  std::vector<ink_thread> threads;
+  for (unsigned int node = 0; node < (unsigned int)caches.size(); ++node) {
+    ink_thread  t;
+    hwloc_obj_t obj = hwloc_get_obj_by_type(ink_get_topology(), HWLOC_OBJ_NODE, node);
+    ink_thread_create(&t, ram_cache_container_thread_init_func, (void *)this, false, 0, nullptr, obj->cpuset);
+    threads.push_back(t);
+  }
+  for (auto t : threads) {
+    ink_thread_join(t);
+  }
+  for (auto cache : caches) {
+    if (!cache) {
+      ink_fatal("Failed to initialize NUMA local ram cache.");
+    }
+  }
+}
+
+constexpr size_t PAGE_SIZE = 4096;
+
+static void *
+align_pointer_to_page(const void *ptr)
+{
+  // Align to the page size (move_page requires that)
+  return reinterpret_cast<void *>(reinterpret_cast<intptr_t>(ptr) & (~(PAGE_SIZE - 1)));
+}
+
+static size_t
+page_count(const void *ptr, size_t size)
+{
+  auto start = reinterpret_cast<intptr_t>(ptr);
+  return (start + size + (PAGE_SIZE - 1)) / PAGE_SIZE - start / PAGE_SIZE;
+}
+
+// returns true if consistent
+static bool
+check_pages_consistency(void *data, size_t size, const char *name = "")
```

Comment:
This function is unused unless `NUMA_CONSISTENCY_CHECK` is defined.  Please add precompile check here.